redis/src/cluster_async/mod.rs

```diff
@@ -499,52 +529,209 @@ where
+                    .await
+                {
+                    Ok(conn) => conn,
+                    Err(_) => random_conn.await,
                 }
             }
-            .boxed()
-            .shared();
-            self.connections
-                .insert(addr.clone(), connection_future.clone());
-            (addr, connection_future)
+        }
+        .boxed()
+        .shared();
+        connections_and_metadata
+            .connections
+            .lock()
+            .await
+            .insert(addr.clone(), connection_future.clone());
+        (addr, connection_future)
+    }
+
+    fn add_future(
+        connections_and_metadata: Arc<ConnectionsAndMetadata<C>>,
+        func: fn(C, Arc<Cmd>) -> RedisFuture<'static, Response>,
+        cmd: Arc<Cmd>,
+        futures: &mut Vec<RedisFuture<Response>>,
+        slot: &SlotAddrs,
+        slot_addr: &SlotAddr,
+    ) {
+        let addr = slot.slot_addr(slot_addr).to_string();
+        futures.push(
+            (async move {
+                let (_, conn) =
+                    Self::get_connection_from_parts(addr, connections_and_metadata).await;
+                let conn = conn.await;
+                func(conn, cmd.clone()).await
+            })
+            .boxed(),
+        );
+    }
+
+    async fn execute_on_all_nodes(
+        func: fn(C, Arc<Cmd>) -> RedisFuture<'static, Response>,
+        cmd: &Arc<Cmd>,
+        only_primaries: bool,
+        connections_and_metadata: Arc<ConnectionsAndMetadata<C>>,
+    ) -> (OperationTarget, RedisResult<Response>) {
+        let operation_count = if only_primaries {
+            connections_and_metadata.connection_count
         } else {
-            // Return a random connection
-            get_random_connection(&self.connections, None)
+            connections_and_metadata.connection_count * 2
+        };
+
+        let mut futures = Vec::with_capacity(operation_count);
+        for slot in connections_and_metadata.slots.values() {
+            Self::add_future(
+                connections_and_metadata.clone(),
+                func,
+                cmd.clone(),
+                &mut futures,
+                slot,
+                &SlotAddr::Master,
+            );
+
+            if !only_primaries {
+                Self::add_future(
+                    connections_and_metadata.clone(),
+                    func,
+                    cmd.clone(),
+                    &mut futures,
+                    slot,
+                    &SlotAddr::Replica,
+                );
+            }
+        }
+        let results = future::join_all(futures).await;
+        let mut merged_results = Vec::with_capacity(operation_count);
+
+        // TODO - we can have better error reporting here if we had an Error variant on Value.
+        for result in results {
+            match result {
+                Ok(response) => match response {
+                    Response::Single(value) => merged_results.push(value),
```

Comment:
does 'value' contains the target node and the result? e.g. (target, res)?

Reply:
no, just the result.

Reply:
So how would the user know which nodes returned the results?

Reply:
they don't. Do other clients return values with sending node?